nav2_ceres_costaware_smoother/include/nav2_ceres_costaware_smoother/smoother.hpp

```diff
@@ -0,0 +1,396 @@
+// Copyright (c) 2021 RoboTech Vision
+// Copyright (c) 2020, Samsung Research America
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License. Reserved.
+
+#ifndef NAV2_CERES_COSTAWARE_SMOOTHER__SMOOTHER_HPP_
+#define NAV2_CERES_COSTAWARE_SMOOTHER__SMOOTHER_HPP_
+
+#include <cmath>
+#include <vector>
+#include <iostream>
+#include <memory>
+#include <queue>
+#include <utility>
+#include <deque>
+#include <limits>
+#include <algorithm>
+
+#include "nav2_ceres_costaware_smoother/smoother_cost_function.hpp"
+#include "nav2_ceres_costaware_smoother/utils.hpp"
+
+#include "ceres/ceres.h"
+#include "Eigen/Core"
+
+namespace nav2_ceres_costaware_smoother
+{
+
+/**
+ * @class nav2_smac_planner::Smoother
+ * @brief A Conjugate Gradient 2D path smoother implementation
+ */
+class Smoother
+{
+public:
+  /**
+   * @brief A constructor for nav2_smac_planner::Smoother
+   */
+  Smoother() {}
+
+  /**
+   * @brief A destructor for nav2_smac_planner::Smoother
+   */
+  ~Smoother() {}
+
+  /**
+   * @brief Initialization of the smoother
+   * @param params OptimizerParam struct
+   */
+  void initialize(const OptimizerParams params)
+  {
+    _debug = params.debug;
+
+    _options.linear_solver_type = ceres::DENSE_QR;
+
+    _options.max_num_iterations = params.max_iterations;
+
+    _options.function_tolerance = params.fn_tol;
+    _options.gradient_tolerance = params.gradient_tol;
+    _options.parameter_tolerance = params.param_tol;
+
+    if (_debug) {
+      _options.minimizer_progress_to_stdout = true;
+      _options.logging_type = ceres::LoggingType::PER_MINIMIZER_ITERATION;
+    } else {
+      _options.logging_type = ceres::SILENT;
+    }
+  }
+
+  /**
+   * @brief Smoother method
+   * @param path Reference to path
+   * @param start_dir Orientation of the first pose
+   * @param end_dir Orientation of the last pose
+   * @param costmap Pointer to minimal costmap
+   * @param params parameters weights
+   * @return If smoothing was successful
+   */
+  bool smooth(
+    std::vector<Eigen::Vector3d> & path,
+    const Eigen::Vector2d & start_dir,
+    const Eigen::Vector2d & end_dir,
+    const nav2_costmap_2d::Costmap2D * costmap,
+    const SmootherParams & params)
+  {
+    // Path has always at least 2 points
+    CHECK(path.size() >= 2) << "Path must have at least 2 points";
```

Comment:
Should we make this `throw` instead? Will the main `nav2_smoother` server catch if the algorithm throws during execution?